File: libs/ontology-kernel/src/domain/media_reference_validator.rs
//! Foundry-style validation for the `media_reference` property type
//! that goes beyond the structural shape check in
//! [`crate::domain::type_system::validate_property_value`].
//!
//! `validate_property_value` is a pure-Rust shape gate. It accepts a
//! string OR an object with a non-empty `uri`/`url` — fine for the
//! pre-H6 stub, but H6 lifts media-reference properties into the
//! Ontology proper and demands two extra invariants from
//! `Using media in the Ontology.md` + `Upload media.md`:
//!
//!   1. **Set existence.** The `mediaSetRid` must address a media set
//!      that exists; an unresolved RID writes a dangling pointer into
//!      object state.
//!   2. **Clearance covers markings.** The user editing the property
//!      must hold every marking on the backing media set
//!      (Foundry: "Adding a media set to your ontology delegates
//!      access control from the media set to the ontology"). The
//!      kernel cannot itself talk to Cedar / Postgres, so we accept
//!      the resolved markings via callback and run the inclusion
//!      check ourselves.
//!
//! Callers (typically `ontology-actions-service` on action submit
//! and the inline-edit path in `handlers/properties.rs`) build a
//! [`MediaReferenceContext`], wire its closures to the platform
//! lookups, and pass it down. The kernel stays dep-light.

use std::collections::BTreeSet;

use serde_json::Value;

/// Canonical Foundry payload for a media-reference property value.
/// Mirrors the shape `core_models::MediaReference` emits via
/// [`MediaReference::to_foundry_json`]; we keep this struct local to
/// the kernel so the kernel does not have to depend on `core-models`.
///
/// The `branch` and `schema` fields are optional in pre-H6 callers
/// (legacy code that wrote string-form pointers); validators MUST
/// surface a clear error if either is missing on H6 surfaces.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ParsedMediaReference {
    pub media_set_rid: String,
    pub media_item_rid: String,
    pub branch: Option<String>,
    pub schema: Option<String>,
}

/// Lookup result for a media set the validator is consulting. The
/// caller resolves this from `media-sets-service` (Postgres) or a
/// cached read; the validator only needs to know whether the set
/// exists and which markings it carries.
#[derive(Debug, Clone)]
pub struct ResolvedMediaSet {
    pub media_set_rid: String,
    /// Lower-cased markings the set is tagged with. Empty = no
    /// markings (the validator allows the edit unconditionally).
    pub markings: Vec<String>,
}

/// Caller-supplied context used by the validator. Two closures —
/// one for set lookup, one for clearance — kept tiny so the action
/// handler can wire them against whichever store it owns without the
/// kernel taking on a new dep.
pub struct MediaReferenceContext<'a> {
    pub resolve_set: Box<dyn Fn(&str) -> Option<ResolvedMediaSet> + Send + Sync + 'a>,
    pub user_clearances: Vec<String>,
}

impl std::fmt::Debug for MediaReferenceContext<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("MediaReferenceContext")
            .field("user_clearances", &self.user_clearances)
            .finish_non_exhaustive()
    }
}

/// Failure modes — distinct variants so the action handler can map
/// each to the right HTTP status without parsing strings.
#[derive(Debug, Clone, thiserror::Error, PartialEq, Eq)]
pub enum MediaReferenceValidationError {
    #[error("media_reference value must be a JSON object on H6 ontology surfaces")]
    NotAnObject,
    #[error("media_reference is missing field `{0}`")]
    MissingField(&'static str),
    #[error("media_reference field `{0}` must be a non-empty string")]
    EmptyField(&'static str),
    #[error("media set `{0}` does not exist")]
    UnknownMediaSet(String),
    #[error("missing clearance: {missing}")]
    InsufficientClearance { missing: String },
}

/// Parse the camelCase Foundry payload (`mediaSetRid`,
/// `mediaItemRid`, …) AND the snake_case Rust payload that
/// `core_models::MediaReference::to_foundry_json` emits. We accept
/// both because the OSDK round-trips through camelCase but the Rust
/// services serialise via the struct.
fn parse(value: &Value) -> Result<ParsedMediaReference, MediaReferenceValidationError> {
    let object = value
        .as_object()
        .ok_or(MediaReferenceValidationError::NotAnObject)?;

    let pull = |camel: &'static str, snake: &'static str| -> Option<&str> {
        object
            .get(camel)
            .or_else(|| object.get(snake))
            .and_then(Value::as_str)
            .map(str::trim)
            .filter(|v| !v.is_empty())
    };

    let media_set_rid = pull("mediaSetRid", "media_set_rid")
        .ok_or(MediaReferenceValidationError::MissingField("mediaSetRid"))?
        .to_string();
    let media_item_rid = pull("mediaItemRid", "media_item_rid")
        .ok_or(MediaReferenceValidationError::MissingField("mediaItemRid"))?
        .to_string();
    let branch = pull("branch", "branch").map(|s| s.to_string());
    let schema = pull("schema", "schema").map(|s| s.to_string());

    Ok(ParsedMediaReference {
        media_set_rid,
        media_item_rid,
        branch,
        schema,
    })
}

/// Run the H6 contract: shape parse → set exists → clearances cover
/// the set's markings. Returns the parsed reference on success so
/// the caller can persist the canonical form (avoids a re-parse).
pub fn validate(
    value: &Value,
    ctx: &MediaReferenceContext<'_>,
) -> Result<ParsedMediaReference, MediaReferenceValidationError> {
    let parsed = parse(value)?;
    let resolved = (ctx.resolve_set)(&parsed.media_set_rid).ok_or_else(|| {
        MediaReferenceValidationError::UnknownMediaSet(parsed.media_set_rid.clone())
    })?;
    if !covers_clearance(&ctx.user_clearances, &resolved.markings) {
        let missing = first_missing_marking(&ctx.user_clearances, &resolved.markings)
            .unwrap_or_else(|| resolved.markings.join(", "));
        return Err(MediaReferenceValidationError::InsufficientClearance { missing });
    }
    Ok(parsed)
}

fn covers_clearance(user_clearances: &[String], required_markings: &[String]) -> bool {
    let user: BTreeSet<&str> = user_clearances
        .iter()
        .map(|s| s.trim())
        .filter(|s| !s.is_empty())
        .collect();
    required_markings
        .iter()
        .map(|s| s.trim())
        .filter(|s| !s.is_empty())
        .all(|marking| user.contains(marking))
}

fn first_missing_marking(user: &[String], required: &[String]) -> Option<String> {
    let user: BTreeSet<&str> = user
        .iter()
        .map(|s| s.trim())
        .filter(|s| !s.is_empty())
        .collect();
    required
        .iter()
        .map(|s| s.trim())
        .find(|marking| !marking.is_empty() && !user.contains(marking))
        .map(|s| s.to_string())
}

/// Build a [`MediaReferenceContext`] from an in-memory map — useful
/// in tests + the inline-edit fast path that already loaded the set.
pub fn context_from_map<'a>(
    sets: std::collections::HashMap<String, ResolvedMediaSet>,
    user_clearances: Vec<String>,
) -> MediaReferenceContext<'a> {
    MediaReferenceContext {
        resolve_set: Box::new(move |rid| sets.get(rid).cloned()),
        user_clearances,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;
    use std::collections::HashMap;

    fn ctx(set_rid: &str, markings: Vec<&str>, clearances: Vec<&str>) -> MediaReferenceContext<'static> {
        let mut map = HashMap::new();
        map.insert(
            set_rid.to_string(),
            ResolvedMediaSet {
                media_set_rid: set_rid.to_string(),
                markings: markings.into_iter().map(String::from).collect(),
            },
        );
        context_from_map(map, clearances.into_iter().map(String::from).collect())
    }

    #[test]
    fn accepts_camel_case_payload() {
        let v = json!({
            "mediaSetRid": "ri.foundry.main.media_set.abc",
            "mediaItemRid": "ri.foundry.main.media_item.def",
            "branch": "main",
            "schema": "IMAGE"
        });
        let ctx = ctx("ri.foundry.main.media_set.abc", vec!["public"], vec!["public"]);
        let parsed = validate(&v, &ctx).expect("happy path");
        assert_eq!(parsed.branch.as_deref(), Some("main"));
        assert_eq!(parsed.schema.as_deref(), Some("IMAGE"));
    }

    #[test]
    fn rejects_unknown_media_set() {
        let v = json!({
            "mediaSetRid": "ri.foundry.main.media_set.absent",
            "mediaItemRid": "x"
        });
        let ctx = ctx("ri.foundry.main.media_set.other", vec![], vec!["public"]);
        let err = validate(&v, &ctx).unwrap_err();
        assert!(matches!(
            err,
            MediaReferenceValidationError::UnknownMediaSet(_)
        ));
    }

    #[test]
    fn rejects_when_clearances_miss_a_marking() {
        let v = json!({
            "mediaSetRid": "ri.foundry.main.media_set.classified",
            "mediaItemRid": "x"
        });
        let ctx = ctx(
            "ri.foundry.main.media_set.classified",
            vec!["public", "secret"],
            vec!["public"],
        );
        let err = validate(&v, &ctx).unwrap_err();
        match err {
            MediaReferenceValidationError::InsufficientClearance { missing } => {
                assert_eq!(missing, "secret")
            }
            other => panic!("expected InsufficientClearance, got {other:?}"),
        }
    }

    #[test]
    fn rejects_missing_required_field() {
        let v = json!({"mediaSetRid": "rid"});
        let ctx = ctx("rid", vec![], vec![]);
        let err = validate(&v, &ctx).unwrap_err();
        assert_eq!(
            err,
            MediaReferenceValidationError::MissingField("mediaItemRid")
        );
    }

    #[test]
    fn rejects_non_object_payload() {
        let v = json!("just-a-string");
        let ctx = ctx("any", vec![], vec![]);
        assert_eq!(
            validate(&v, &ctx).unwrap_err(),
            MediaReferenceValidationError::NotAnObject
        );
    }

    #[test]
    fn accepts_snake_case_payload_too() {
        let v = json!({
            "media_set_rid": "ri.foundry.main.media_set.abc",
            "media_item_rid": "ri.foundry.main.media_item.def"
        });
        let ctx = ctx("ri.foundry.main.media_set.abc", vec![], vec![]);
        validate(&v, &ctx).unwrap();
    }
}

File: libs/ontology-kernel/domain/media_reference_validator.go
// Foundry-style validation for the `media_reference` property type
// that goes beyond the structural shape check in
// [ValidatePropertyValue] (type_system.go).
//
// `ValidatePropertyValue` is a pure JSON shape gate. It accepts a
// string OR an object with a non-empty `uri`/`url` — fine for the
// pre-H6 stub. H6 lifts media-reference properties into the Ontology
// proper and demands two extra invariants:
//
//  1. **Set existence.** The mediaSetRid must address a media set
//     that exists.
//  2. **Clearance covers markings.** The user editing the property
//     must hold every marking on the backing media set.
//
// Mirrors `libs/ontology-kernel/src/domain/media_reference_validator.rs`.

package domain

import (
	"encoding/json"
	"errors"
	"fmt"
	"sort"
	"strings"
)

// ParsedMediaReference mirrors `struct ParsedMediaReference`.
type ParsedMediaReference struct {
	MediaSetRID  string
	MediaItemRID string
	Branch       *string
	Schema       *string
}

// ResolvedMediaSet mirrors `struct ResolvedMediaSet`. Markings are
// expected lower-cased by callers (the Rust source documents the same
// contract).
type ResolvedMediaSet struct {
	MediaSetRID string
	Markings    []string
}

// MediaReferenceContext mirrors `struct MediaReferenceContext<'a>`.
// `ResolveSet` is the closure the Rust source carries as
// `Box<dyn Fn(&str) -> Option<ResolvedMediaSet>>`; the Go equivalent
// is a function value plus the user's clearances.
type MediaReferenceContext struct {
	ResolveSet     func(rid string) *ResolvedMediaSet
	UserClearances []string
}

// MediaReferenceValidationErrorKind tags the error variants. Mirrors
// `enum MediaReferenceValidationError` in the Rust source.
type MediaReferenceValidationErrorKind int

const (
	// MediaRefNotAnObject — payload is not a JSON object.
	MediaRefNotAnObject MediaReferenceValidationErrorKind = iota
	// MediaRefMissingField — required field is absent.
	MediaRefMissingField
	// MediaRefEmptyField — required field is present but blank.
	MediaRefEmptyField
	// MediaRefUnknownMediaSet — resolveSet returned nil.
	MediaRefUnknownMediaSet
	// MediaRefInsufficientClearance — user is missing one of the
	// markings required by the resolved media set.
	MediaRefInsufficientClearance
)

// MediaReferenceValidationError mirrors the variant enum. Field
// presence depends on Kind: MissingField/EmptyField → Field;
// UnknownMediaSet → MediaSetRID; InsufficientClearance → Missing.
type MediaReferenceValidationError struct {
	Kind        MediaReferenceValidationErrorKind
	Field       string
	MediaSetRID string
	Missing     string
}

// Error mirrors the Display impl — produces the same human-readable
// string the Rust `thiserror` macro generates.
func (e *MediaReferenceValidationError) Error() string {
	switch e.Kind {
	case MediaRefNotAnObject:
		return "media_reference value must be a JSON object on H6 ontology surfaces"
	case MediaRefMissingField:
		return fmt.Sprintf("media_reference is missing field `%s`", e.Field)
	case MediaRefEmptyField:
		return fmt.Sprintf("media_reference field `%s` must be a non-empty string", e.Field)
	case MediaRefUnknownMediaSet:
		return fmt.Sprintf("media set `%s` does not exist", e.MediaSetRID)
	case MediaRefInsufficientClearance:
		return fmt.Sprintf("missing clearance: %s", e.Missing)
	}
	return "media_reference validation error"
}

var _ error = (*MediaReferenceValidationError)(nil)

// IsMediaRefError reports whether the error chain contains a
// [MediaReferenceValidationError] of the given kind.
func IsMediaRefError(err error, kind MediaReferenceValidationErrorKind) bool {
	var m *MediaReferenceValidationError
	if !errors.As(err, &m) {
		return false
	}
	return m.Kind == kind
}

// parseMediaReference mirrors `fn parse`. Accepts both camelCase
// (`mediaSetRid`, `mediaItemRid`, …) and snake_case
// (`media_set_rid`, `media_item_rid`, …) payloads — the OSDK
// round-trips through camelCase but the Rust services serialise via
// the struct, so both must work.
func parseMediaReference(value json.RawMessage) (ParsedMediaReference, error) {
	var obj map[string]json.RawMessage
	if err := json.Unmarshal(value, &obj); err != nil || obj == nil {
		return ParsedMediaReference{}, &MediaReferenceValidationError{Kind: MediaRefNotAnObject}
	}

	pull := func(camel, snake string) (string, bool) {
		raw, ok := obj[camel]
		if !ok {
			raw, ok = obj[snake]
			if !ok {
				return "", false
			}
		}
		var s string
		if err := json.Unmarshal(raw, &s); err != nil {
			return "", false
		}
		s = strings.TrimSpace(s)
		if s == "" {
			return "", false
		}
		return s, true
	}

	mediaSetRID, ok := pull("mediaSetRid", "media_set_rid")
	if !ok {
		return ParsedMediaReference{}, &MediaReferenceValidationError{
			Kind: MediaRefMissingField, Field: "mediaSetRid",
		}
	}
	mediaItemRID, ok := pull("mediaItemRid", "media_item_rid")
	if !ok {
		return ParsedMediaReference{}, &MediaReferenceValidationError{
			Kind: MediaRefMissingField, Field: "mediaItemRid",
		}
	}

	parsed := ParsedMediaReference{
		MediaSetRID:  mediaSetRID,
		MediaItemRID: mediaItemRID,
	}
	if v, ok := pull("branch", "branch"); ok {
		parsed.Branch = &v
	}
	if v, ok := pull("schema", "schema"); ok {
		parsed.Schema = &v
	}
	return parsed, nil
}

// ValidateMediaReference mirrors `pub fn validate`. Runs the H6
// contract: shape parse → set exists → clearances cover the set's
// markings. Returns the parsed reference on success so the caller
// can persist the canonical form (avoids a re-parse).
func ValidateMediaReference(value json.RawMessage, ctx MediaReferenceContext) (ParsedMediaReference, error) {
	parsed, err := parseMediaReference(value)
	if err != nil {
		return ParsedMediaReference{}, err
	}
	if ctx.ResolveSet == nil {
		return ParsedMediaReference{}, &MediaReferenceValidationError{
			Kind: MediaRefUnknownMediaSet, MediaSetRID: parsed.MediaSetRID,
		}
	}
	resolved := ctx.ResolveSet(parsed.MediaSetRID)
	if resolved == nil {
		return ParsedMediaReference{}, &MediaReferenceValidationError{
			Kind: MediaRefUnknownMediaSet, MediaSetRID: parsed.MediaSetRID,
		}
	}
	if !coversClearance(ctx.UserClearances, resolved.Markings) {
		missing := firstMissingMarking(ctx.UserClearances, resolved.Markings)
		if missing == "" {
			missing = strings.Join(resolved.Markings, ", ")
		}
		return ParsedMediaReference{}, &MediaReferenceValidationError{
			Kind: MediaRefInsufficientClearance, Missing: missing,
		}
	}
	return parsed, nil
}

// coversClearance mirrors `fn covers_clearance`. Builds a sorted
// set of the user's non-blank clearances and verifies every
// non-blank required marking lives in it.
func coversClearance(userClearances, requiredMarkings []string) bool {
	user := stringSet(userClearances)
	for _, marking := range requiredMarkings {
		marking = strings.TrimSpace(marking)
		if marking == "" {
			continue
		}
		if !user[marking] {
			return false
		}
	}
	return true
}

// firstMissingMarking mirrors `fn first_missing_marking`. Iteration
// order matches Rust: the required slice's order is preserved.
func firstMissingMarking(userClearances, requiredMarkings []string) string {
	user := stringSet(userClearances)
	for _, marking := range requiredMarkings {
		marking = strings.TrimSpace(marking)
		if marking == "" {
			continue
		}
		if !user[marking] {
			return marking
		}
	}
	return ""
}

// stringSet builds a set from a slice, trimming + dropping blanks.
// Mirrors the BTreeSet construction in the Rust source.
func stringSet(values []string) map[string]bool {
	out := map[string]bool{}
	for _, v := range values {
		v = strings.TrimSpace(v)
		if v == "" {
			continue
		}
		out[v] = true
	}
	return out
}

// MediaReferenceContextFromMap mirrors `pub fn context_from_map`.
// Useful in tests + the inline-edit fast path that already loaded the
// set.
func MediaReferenceContextFromMap(sets map[string]ResolvedMediaSet, userClearances []string) MediaReferenceContext {
	// Snapshot keys so the closure's view of the map is stable even
	// if the caller mutates it after building the context.
	keys := make([]string, 0, len(sets))
	for k := range sets {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	frozen := make(map[string]ResolvedMediaSet, len(sets))
	for _, k := range keys {
		frozen[k] = sets[k]
	}
	return MediaReferenceContext{
		ResolveSet: func(rid string) *ResolvedMediaSet {
			if v, ok := frozen[rid]; ok {
				return &v
			}
			return nil
		},
		UserClearances: append([]string(nil), userClearances...),
	}
}
